### PR TITLE
perf: cache l2 block range

### PR DESCRIPTION
**Pull request by brtkx**
Implements mutex lock to cache l2BlockRange

### CODE REVIEW

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 1]
```diff
@@ -61,6 +61,8 @@ const ASSERTION_CREATED_PADDING = 50
 // expected number of L1 blocks that it takes for a validator to confirm an L1 block after the node deadline is passed
 const ASSERTION_CONFIRMED_PADDING = 20
 
+const l2BlockRangeCache: { [key in number]: number[] } = {}
```
Comment (spsjvc) [inline src/lib/message/L2ToL1MessageNitro.ts:64]: This needs to also be grouped by parent chain

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 2]
```diff
@@ -155,6 +157,10 @@ export class L2ToL1MessageReaderNitro extends L2ToL1MessageNitro {
     return outbox.callStatic.isSpent(this.event.position)
   }
 
+  public setL2BlockRangeCache(l1Block: number, l2BlockRange: number[]) {
+    l2BlockRangeCache[l1Block] = l2BlockRange
+  }
```
Comment (spsjvc) [inline src/lib/message/L2ToL1MessageNitro.ts:162]: I would remove this method and just directly update the cache above from the code

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 3]
```diff
@@ -212,17 +218,24 @@ export class L2ToL1MessageReaderNitro extends L2ToL1MessageNitro {
...
         if (!startBlock || !endBlock) {
           throw new Error()
         }
         createdFromBlock = BigNumber.from(startBlock)
         createdToBlock = BigNumber.from(endBlock)
+
+        this.setL2BlockRangeCache(createdAtBlock.toNumber(), [
+          startBlock,
+          endBlock,
+        ])
```
Comment (spsjvc) [inline src/lib/message/L2ToL1MessageNitro.ts:238]: Maybe we should cache it below the try/catch

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 4]
```diff
@@ -212,17 +218,24 @@ export class L2ToL1MessageReaderNitro extends L2ToL1MessageNitro {
...
-        const l2BlockRange = await getBlockRangesForL1Block({
-          forL1Block: createdAtBlock.toNumber(),
-          provider: this.l1Provider as JsonRpcProvider,
-        })
+        const l2BlockRange =
+          l2BlockRangeCache?.[createdAtBlock.toNumber()] ||
+          (await getBlockRangesForL1Block({
+            forL1Block: createdAtBlock.toNumber(),
+            provider: this.l1Provider as JsonRpcProvider,
+          }))
```
Comment (spsjvc) [inline src/lib/message/L2ToL1MessageNitro.ts:226]: nit for readability

```suggestion
        const cachedL2BlockRange = l2BlockRangeCache?.[createdAtBlock.toNumber()]
        
        return cachedL2BlockRange ?? (await getBlockRangesForL1Block({
          forL1Block: createdAtBlock.toNumber(),
          provider: this.l1Provider as JsonRpcProvider,
        }))
```

File: src/lib/message/L2ToL1MessageNitro.ts (hunk 1 above)
Comment (spsjvc) [reply src/lib/message/L2ToL1MessageNitro.ts:64]: My bad, needs to be grouped by chain, cause one parent chain (e.g. Arb Goerli) can have many Orbit chains

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 5]
```diff
@@ -211,11 +231,21 @@ export class L2ToL1MessageReaderNitro extends L2ToL1MessageNitro {
...
+        chainId,
+        l1BlockNumber: createdAtBlock.toNumber(),
+      })
+
       try {
-        const l2BlockRange = await getBlockRangesForL1Block({
-          forL1Block: createdAtBlock.toNumber(),
-          provider: this.l1Provider as JsonRpcProvider,
-        })
+        const cachedL2BlockRange = getL2BlockRangeCache(l2BlockRangeCacheKey)
```
Comment (gzeoneth) [inline src/lib/message/L2ToL1MessageNitro.ts:242]: we should acquire a lock here on cache miss to block parallel call

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 6]
```diff
@@ -228,6 +258,10 @@ export class L2ToL1MessageReaderNitro extends L2ToL1MessageNitro {
         createdFromBlock = createdAtBlock
         createdToBlock = createdAtBlock
       }
+
+      if (l2BlockRange) {
+        setL2BlockRangeCache(l2BlockRangeCacheKey, l2BlockRange)
```
Comment (gzeoneth) [inline src/lib/message/L2ToL1MessageNitro.ts:263]: then release the lock here

**Review by gzeoneth — COMMENTED** [review]
While caching `(chainId, l1BlockNumber) -> l2BlockRange` can save around 50% of the rpc call for unconfirmed node (since we always call `getBlockFromNodeNum` on latest confirmed), a better way seems to be caching the `(chainId, blockNumber) -> l1BlockNumber` result. 

By using a bitwise binary search, it should be possible to save more than 75% rpc calls for unconfirmed node (50% from the latest confirmed, another 50%+ because of the duplicated binary search in `getBlockRangesForL1Block`) and the fact that those l2 blocks are nearby

A bitwise binary search improve cache hit since you will have the same midpoint assuming the l2 blocks are in nearby range (which is always `>latestConfirmed`).

**Review by gzeoneth — APPROVED** [review]
LGTM

File: src/lib/message/L2ToL1MessageNitro.ts [hunk 7]
```diff
@@ -61,6 +62,64 @@ const ASSERTION_CREATED_PADDING = 50
...
+
+async function getL2BlockRangeCache({
+  l1Provider,
+  l2Provider,
+  l1BlockNumber,
+}: {
+  l1Provider: JsonRpcProvider
+  l2Provider: JsonRpcProvider
+  l1BlockNumber: number
+}) {
```
Comment (spsjvc) [inline src/lib/message/L2ToL1MessageNitro.ts:90]:
```suggestion
async function getBlockRangesForL1BlockWithCache({
  l1Provider,
  l2Provider,
  forL1Block,
}: {
  l1Provider: JsonRpcProvider
  l2Provider: JsonRpcProvider
  forL1Block: number
}) {
```

**Review by spsjvc — APPROVED** [review]

**Pull request merged**